libifupdown/waitif_libmnl.c

```diff
@@ -0,0 +1,126 @@
+#include <time.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <pthread.h>
+#include <sys/types.h>
+
+#include <net/if.h>
+#include <libmnl/libmnl.h>
+#include <linux/if.h>
+#include <linux/if_link.h>
+#include <linux/rtnetlink.h>
+
+#include "libifupdown/dict.h"
+#include "libifupdown/waitif.h"
+
+struct waitif_listener {
+	pthread_t pthread;
+	struct mnl_socket *nl;
+	struct lif_dict ifs;
+	pthread_mutex_t ifs_mtx;
+};
+
+/* Only one listener will be needed for ifupdown, thus global. */
+static struct waitif_listener listener;
+
+static int
+netlink_cb(const struct nlmsghdr *nlh, void *arg)
+{
+	struct waitif_listener *ln = (struct waitif_listener *)arg;
+
+	struct ifinfomsg *ifm;
+	ifm = mnl_nlmsg_get_payload(nlh);
+
+	char ifname[IF_NAMESIZE];
+	if (!if_indextoname(ifm->ifi_index, ifname)) {
+		fprintf(stderr, "ifupdown: if_indextoname failed for interface %d: %s\n", ifm->ifi_index, strerror(errno));
+		return MNL_CB_OK; /* continue processing other ifaces */
+	}
+
+	pthread_mutex_lock(&ln->ifs_mtx);
+	struct lif_dict_entry *entry;
+	if (!(entry = lif_dict_find(&ln->ifs, ifname))) {
+		pthread_mutex_unlock(&ln->ifs_mtx);
+		return MNL_CB_OK; /* no link state handling requested */
+	}
+
+	struct waitif_iface *iface;
+	iface = (struct waitif_iface *)entry->data;
+	if (ifm->ifi_flags & iface->target_state) {
+		sem_post(&iface->sema);
+		lif_dict_delete(&ln->ifs, ifname);
+	}
+	pthread_mutex_unlock(&ln->ifs_mtx);
+
+	return MNL_CB_OK;
+}
+
+static void *
+netlink_loop(void *arg)
+{
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+	struct waitif_listener *ln = (struct waitif_listener*)arg;
+
+	ssize_t ret = mnl_socket_recvfrom(listener.nl, buf, sizeof(buf));
+	while (ret > 0) {
+		ret = mnl_cb_run(buf, (size_t)ret, 0, 0, netlink_cb, (void *)ln);
+		if (ret <= 0)
+			break;
+		ret = mnl_socket_recvfrom(listener.nl, buf, sizeof(buf));
+	}
+	if (ret == -1)
+		fprintf(stderr, "ifupdown: netlink_loop failed %s\n", strerror(errno));
+
+	return NULL;
+}
+
+bool
+lif_waitif_init(void)
+{
+	lif_dict_init(&listener.ifs);
+	if (pthread_mutex_init(&listener.ifs_mtx, NULL))
```

Comment:
I would prefer to avoid threads for this.  Why not have a child process?

Reply:
I actually abandoned this approach (integrating link state detection directly into ifupdown-ng) in favor of using an executor-based approach, based on the feedback in https://github.com/ifupdown-ng/ifupdown-ng/issues/178#issuecomment-1153054514

The current implementation of the executor is in https://github.com/nmeum/ifupdown-ng-waitif/blob/main/waitif.c I can send a PR to integrate that if there is an interest in this. The executor-based approach also uses pthreads presently simply because communicating with a thread is easier then doing IPC. However, I can also rewrite it to `fork()` instead.

For the executor-based approach it would also be nice to merge #179.